<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Function reference â€¢ qfm</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">qfm</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/basic_tutorial.html">Basic tutorial on Quantitative Fate Mapping (QFM)</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Function reference</h1>
    </div>

    <div class="section level2">
      <h2 id="all-functions">All functions<a class="anchor" aria-label="anchor" href="#all-functions"></a></h2>
      
      

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="add_edge_length.html">add_edge_length()</a></code> 
        </dt>
        <dd>Add edge length to edge df</dd>
      </dl><dl><dt>
          
          <code><a href="as.phylo.type_graph.html">as.phylo(<i>&lt;type_graph&gt;</i>)</a></code> 
        </dt>
        <dd>Convert type graph to phylo object</dd>
      </dl><dl><dt>
          
          <code><a href="calc_bsum.html">calc_bsum()</a></code> 
        </dt>
        <dd>Compute BSUM or Colless index value</dd>
      </dl><dl><dt>
          
          <code><a href="calculate_type_counts.html">calculate_type_counts()</a></code> 
        </dt>
        <dd>Calculate the counts before next division</dd>
      </dl><dl><dt>
          
          <code><a href="cells_to_edges.html">cells_to_edges()</a></code> 
        </dt>
        <dd>Convert cells to edge data.frame</dd>
      </dl><dl><dt>
          
          <code><a href="cells_to_nodes.html">cells_to_nodes()</a></code> 
        </dt>
        <dd>Convert cells to node data.frame</dd>
      </dl><dl><dt>
          
          <code><a href="change_target_time.html">change_target_time()</a></code> 
        </dt>
        <dd>Change target time</dd>
      </dl><dl><dt>
          
          <code><a href="check_n_identity.html">check_n_identity()</a></code> 
        </dt>
        <dd>Compre if two strings are identical other than 'N' characters</dd>
      </dl><dl><dt>
          
          <code><a href="collaspe_di.html">collaspe_di()</a></code> 
        </dt>
        <dd>collaspe nodes below certain threshold</dd>
      </dl><dl><dt>
          
          <code><a href="concat_mut_p.html">concat_mut_p()</a></code> 
        </dt>
        <dd>Concatenate two `mut_p` (mutagenesis parameters) objects.</dd>
      </dl><dl><dt>
          
          <code><a href="concat_mut_p_list.html">concat_mut_p_list()</a></code> 
        </dt>
        <dd>Concatenate a list of `mut_p` (mutagenesis parameters) objects.</dd>
      </dl><dl><dt>
          
          <code><a href="construct_true_lineage_multi.html">construct_true_lineage_multi()</a></code> 
        </dt>
        <dd>deprecated now (repplaced by mod2)
deprecated but no mod2 available now</dd>
      </dl><dl><dt>
          
          <code><a href="distribute_all_mut_counts.html">distribute_all_mut_counts()</a></code> 
        </dt>
        <dd>Distribute allele counts for all elements</dd>
      </dl><dl><dt>
          
          <code><a href="distribute_mut_counts.html">distribute_mut_counts()</a></code> 
        </dt>
        <dd>Distribute mutant allele counts between two groups</dd>
      </dl><dl><dt>
          
          <code><a href="edges_df_to_newick.html">edges_df_to_newick()</a></code> 
        </dt>
        <dd>Convert simplified edges df to phylo</dd>
      </dl><dl><dt>
          
          <code><a href="generate_diff_outcome.html">generate_diff_outcome()</a></code> 
        </dt>
        <dd>Generate differentiation outcome matrix for type graph</dd>
      </dl><dl><dt>
          
          <code><a href="generate_edge_df.html">generate_edge_df()</a></code> 
        </dt>
        <dd>Generate edge data.frame for type graph, getting edge lengths from laid out graph</dd>
      </dl><dl><dt>
          
          <code><a href="generate_mut_ids.html">generate_mut_ids()</a></code> 
        </dt>
        <dd>Generate a batch of depth-1 mutante allele</dd>
      </dl><dl><dt>
          
          <code><a href="generate_nonrecur_mut_ids.html">generate_nonrecur_mut_ids()</a></code> 
        </dt>
        <dd>Generate a batch of non recurring depth-1 mutante allele</dd>
      </dl><dl><dt>
          
          <code><a href="generate_recur_mut_ids.html">generate_recur_mut_ids()</a></code> 
        </dt>
        <dd>Generate a batch of recurring depth-1 mutant allele</dd>
      </dl><dl><dt>
          
          <code><a href="get_cell_type_counts.html">get_cell_type_counts()</a></code> 
        </dt>
        <dd>current version of cell phylogeny model</dd>
      </dl><dl><dt>
          
          <code><a href="get_edge_diff.html">get_edge_diff()</a></code> 
        </dt>
        <dd>ICE-FASE inference when topology has multifurcations</dd>
      </dl><dl><dt>
          
          <code><a href="get_edge_len.html">get_edge_len()</a></code> 
        </dt>
        <dd>get edge length statistics to distinguish between g0 and g1</dd>
      </dl><dl><dt>
          
          <code><a href="get_mut_duration.html">get_mut_duration()</a></code> 
        </dt>
        <dd>get overlaps of intervals using trick on IRanges
TODO: dependes on IRanges</dd>
      </dl><dl><dt>
          
          <code><a href="get_next.html">get_next()</a></code> 
        </dt>
        <dd>extend out node to the next bifurcation in the lineage tree</dd>
      </dl><dl><dt>
          
          <code><a href="get_node_mapping.html">get_node_mapping()</a></code> 
        </dt>
        <dd>get mapping of inferred progenitor states to true progenitor states</dd>
      </dl><dl><dt>
          
          <code><a href="get_sample_size.html">get_sample_size()</a></code> 
        </dt>
        <dd>Get sample size of sim_history</dd>
      </dl><dl><dt>
          
          <code><a href="get_sampled_cells.html">get_sampled_cells()</a></code> 
        </dt>
        <dd>Get sampled cells of at target time</dd>
      </dl><dl><dt>
          
          <code><a href="get_t_total.html">get_t_total()</a></code> 
        </dt>
        <dd>deprecated function, used for somatic mutation
get the total amount of time passed that the site could have mutated</dd>
      </dl><dl><dt>
          
          <code><a href="get_target_time.html">get_target_time()</a></code> 
        </dt>
        <dd>Get target time from sim_history</dd>
      </dl><dl><dt>
          
          <code><a href="get_true_size_mod2.html">get_true_size_mod2()</a></code> 
        </dt>
        <dd>this function returns the incoming size as well as the outgoing sizes</dd>
      </dl><dl><dt>
          
          <code><a href="hello.html">hello()</a></code> 
        </dt>
        <dd>Hello, World!</dd>
      </dl><dl><dt>
          
          <code><a href="ice_fase.html">ice_fase()</a></code> 
        </dt>
        <dd>reconstructs phylogeny with phylotime, and infers quantitative fate map with ice_fase</dd>
      </dl><dl><dt>
          
          <code><a href="impute_characters.html">impute_characters()</a></code> 
        </dt>
        <dd>Impute missing characters from cell allele matrix</dd>
      </dl><dl><dt>
          
          <code><a href="init_mut_counts_list.html">init_mut_counts_list()</a></code> 
        </dt>
        <dd>Initialize mutant allele count list for bulk counting</dd>
      </dl><dl><dt>
          
          <code><a href="initial_states_list.html">initial_states_list()</a></code> 
        </dt>
        <dd>Make initial states list</dd>
      </dl><dl><dt>
          
          <code><a href="initialize_mut_element_counter.html">initialize_mut_element_counter()</a></code> 
        </dt>
        <dd>Initialize the global mut_element counter for non-recurring alleles</dd>
      </dl><dl><dt>
          
          <code><a href="initialize_mut_state.html">initialize_mut_state()</a></code> 
        </dt>
        <dd>Initialize unmutated matrix for single cell model</dd>
      </dl><dl><dt>
          
          <code><a href="is_non_decs.html">is_non_decs()</a></code> 
        </dt>
        <dd>Check if a vector is non-decreasing</dd>
      </dl><dl><dt>
          
          <code><a href="is_zero_range.html">is_zero_range()</a></code> 
        </dt>
        <dd>Check if a vector is all the same value</dd>
      </dl><dl><dt>
          
          <code><a href="make_gr_tr_data.html">make_gr_tr_data()</a></code> 
        </dt>
        <dd>mostly deprecated functions of ICE_FASE, replaced by 'ice_fase_mod1.R'
some functions are still used for processing downstream results</dd>
      </dl><dl><dt>
          
          <code><a href="make_heatmap_col.html">make_heatmap_col()</a></code> 
        </dt>
        <dd>Make a color palette for Heatmap visualization</dd>
      </dl><dl><dt>
          
          <code><a href="make_mut_element.html">make_mut_element()</a></code> 
        </dt>
        <dd>Constructor for mutation parameter for a single element
Only supports max_depth of 1</dd>
      </dl><dl><dt>
          
          <code><a href="make_mut_param_by_rate.html">make_mut_param_by_rate()</a></code> 
        </dt>
        <dd>Make a mut_param by a sequence of mutation rates</dd>
      </dl><dl><dt>
          
          <code><a href="make_mut_param_by_rate_rvec.html">make_mut_param_by_rate_rvec()</a></code> 
        </dt>
        <dd>Make a mut_param by a sequence of mutation rates and multinomial allele probabilities</dd>
      </dl><dl><dt>
          
          <code><a href="make_type_graph.html">make_type_graph()</a></code> 
        </dt>
        <dd>Constructor for type graph</dd>
      </dl><dl><dt>
          
          <code><a href="mut_counts_total.html">mut_counts_total()</a></code> 
        </dt>
        <dd>Gives the counts for mut_counts_list</dd>
      </dl><dl><dt>
          
          <code><a href="mutate_barcodes.html">mutate_barcodes()</a></code> 
        </dt>
        <dd>Generate mutant allele for single cell mutation states</dd>
      </dl><dl><dt>
          
          <code><a href="mutate_cells.html">mutate_cells()</a></code> 
        </dt>
        <dd>Mutate the barcodes for cells object</dd>
      </dl><dl><dt>
          
          <code><a href="normalize_row.html">normalize_row()</a></code> 
        </dt>
        <dd>Normalize rows of a matrix to be the same sum `m`</dd>
      </dl><dl><dt>
          
          <code><a href="normalize_tree.html">normalize_tree()</a></code> 
        </dt>
        <dd>Construct lineage tree from sim_history</dd>
      </dl><dl><dt>
          
          <code><a href="num_gen_single_type.html">num_gen_single_type()</a></code> 
        </dt>
        <dd>Given a single cell type, generate the number of generations it will divide until it reaches target time or differentiate</dd>
      </dl><dl><dt>
          
          <code><a href="phylotime.html">phylotime()</a></code> 
        </dt>
        <dd>Phylotime for reconstructing time-scaled phylogeny based on lineage barcodes</dd>
      </dl><dl><dt>
          
          <code><a href="plot_col_vec.html">plot_col_vec()</a></code> 
        </dt>
        <dd>Plot a color vector for visualization</dd>
      </dl><dl><dt>
          
          <code><a href="plot_gr_data.html">plot_gr_data()</a></code> 
        </dt>
        <dd>plot ice_fase results</dd>
      </dl><dl><dt>
          
          <code><a href="plot_legend.html">plot_legend()</a></code> 
        </dt>
        <dd>Plot legend on a new canvas given a ggplot object</dd>
      </dl><dl><dt>
          
          <code><a href="plot_type_graph.html">plot_type_graph()</a></code> 
        </dt>
        <dd>Plot type graph using ggraph</dd>
      </dl><dl><dt>
          
          <code><a href="push_pdf.html">push_pdf()</a></code> 
        </dt>
        <dd>Push a ggplot object to a PDF file</dd>
      </dl><dl><dt>
          
          <code><a href="push_png.html">push_png()</a></code> 
        </dt>
        <dd>Push a ggplot object to a PNG file</dd>
      </dl><dl><dt>
          
          <code><a href="rstoc.html">rstoc()</a></code> 
        </dt>
        <dd>Round a number stochasticly, returns integer greater than or equal to one.</dd>
      </dl><dl><dt>
          
          <code><a href="sample_all_mut_counts.html">sample_all_mut_counts()</a></code> 
        </dt>
        <dd>Draw one sample of mutation counts for all elements</dd>
      </dl><dl><dt>
          
          <code><a href="sample_all_mut_counts_gen.html">sample_all_mut_counts_gen()</a></code> 
        </dt>
        <dd>Draw one sample of the mutation counts after a number of generatioins</dd>
      </dl><dl><dt>
          
          <code><a href="sample_doublet_branch.html">sample_doublet_branch()</a></code> 
        </dt>
        <dd>the stochastic coal function
sample the number of doublet branches when dividing n cells into 2*n cell</dd>
      </dl><dl><dt>
          
          <code><a href="sample_doublet_twotype.html">sample_doublet_twotype()</a></code> 
        </dt>
        <dd>the asymmetric coal function</dd>
      </dl><dl><dt>
          
          <code><a href="sample_mut_counts.html">sample_mut_counts()</a></code> 
        </dt>
        <dd>Draw one sample of the mutation counts after some time</dd>
      </dl><dl><dt>
          
          <code><a href="scale_counts.html">scale_counts()</a></code> 
        </dt>
        <dd>deprecated
Assign counts based on total and fraction, rounded to integer
Ensure that each category gets some minimum of counts
Scale counts based on a factor, rounded to integer, counts must be at least 1</dd>
      </dl><dl><dt>
          
          <code><a href="sim_history_to_edges.html">sim_history_to_edges()</a></code> 
        </dt>
        <dd>Convert sim_history to edges data.frame</dd>
      </dl><dl><dt>
          
          <code><a href="sim_history_to_nodes.html">sim_history_to_nodes()</a></code> 
        </dt>
        <dd>Convert sim_history to nodes data.frame</dd>
      </dl><dl><dt>
          
          <code><a href="simplify_edges_df.html">simplify_edges_df()</a></code> 
        </dt>
        <dd>Remove internal nodes with single daughter</dd>
      </dl><dl><dt>
          
          <code><a href="subsample_all_mut_counts.html">subsample_all_mut_counts()</a></code> 
        </dt>
        <dd>Draw subsample of allele counts for all elements</dd>
      </dl><dl><dt>
          
          <code><a href="subsample_mut_counts.html">subsample_mut_counts()</a></code> 
        </dt>
        <dd>Draw a subsample of mutant allele counts</dd>
      </dl><dl><dt>
          
          <code><a href="subset_mut_p.html">subset_mut_p()</a></code> 
        </dt>
        <dd>Subset a `mut_p` object to a number of its elements</dd>
      </dl><dl><dt>
          
          <code><a href="white_col_mapper.html">white_col_mapper()</a></code> 
        </dt>
        <dd>Generate a 'white' color vector for plotting</dd>
      </dl></div>
  </main></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Weixiang Fang.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer></div>

  

  

  </body></html>

